### PR TITLE
Commerzbank - dividends

### PR DESCRIPTION
**Pull request by oxi0**
Entered new logics for dividends and added testcases.

This solves #824 #827 #915

I hope I've done everything right so far.

### CODE REVIEW

**Comment by oxi0** [comment]
@SirGibihm I noticed that the functions in the sell area all uses the buy functions. Should the function names be adapted to BuySell as in the last pull request?

**Comment by SirGibihm** [comment]
> 
> 
> @SirGibihm I noticed that the functions in the sell area all uses the buy functions. Should the function names be adapted to BuySell as in the last pull request?

Combining functions does make sense as long as it reduces code duplication and code clutter. So yeah, the make it BuySells

File: src/brokers/commerzbank.js [hunk 1]
```diff
@@ -46,12 +49,23 @@ const findTimeSell = textArr =>
...
+	  return textArr[textArr.findIndex(t => t.includes('Information')) - 3];
+	}
+  }
+  if (textArr.indexOf('Valuta') > 0) {
+	if (textArr[textArr.findIndex(t => t.includes('Valuta')) + 1].split(".").length > 1) {
+      return textArr[textArr.findIndex(t => t.includes('Valuta')) + 1]
+	} 
+  }
+  if (textArr.indexOf('per') > 0) {
+	if (textArr[textArr.findIndex(t => t.includes('per')) + 1].split(".").length > 1) {
```
Comment (dredav) [inline src/brokers/commerzbank.js:62]: Did you run the eslint pre-commit hock at commit? The hock will fix this automatically

```suggestion
	if (textArr[textArr.findIndex(t => t.includes('per')) + 1].split('.').length > 1) {
```

File: src/brokers/commerzbank.js [hunk 2]
```diff
@@ -65,19 +79,32 @@ const findIsinDividend = (textArr, foreignDividend) => {
   if (foreignDividend) {
     isinIdx = textArr.findIndex(t => t.includes('STK')) + 2;
   } else {
-    isinIdx = textArr.findIndex(t => t.includes('ISIN')) + 3;
+    if (textArr.indexOf('WKN/ISIN') > 0) {
+	  if (textArr.indexOf('Inhaber-Anteile') > 0) {
+	    isinIdx = textArr.findIndex(t => t.includes('Inhaber-Anteile')) - 1;
```
Comment (dredav) [inline src/brokers/commerzbank.js:84]: So you are checking `indexOf('Inhaber-Anteile')` and then searching for a line that is not equal `Inhaber-Anteile` and only contains `Inhaber-Anteile`. Whats abot this pseudo code:
cost lineNumber = textArr.findIndex(t => t.includes('Inhaber-Anteile'));
if (lineNumber >= 0) { isinIdx = lineNumber - 1}

Comment (dredav) [reply src/brokers/commerzbank.js:84]: Same thing at L85-86

File: src/brokers/commerzbank.js [hunk 3]
```diff
@@ -65,19 +79,32 @@ const findIsinDividend = (textArr, foreignDividend) => {
...
   const priceIdx = textArr.findIndex(t => t.includes('Kurswert')) - 1;
   return parseGermanNum(textArr[priceIdx]);
 };
 
-const findFeeBuy = (textArr, amount) => {
-  const payedAmountIdx = textArr.findIndex(t => t.includes('Valuta')) + 15;
-  return +Big(parseGermanNum(textArr[payedAmountIdx])).minus(amount);
+const findFeeBuySell = (textArr, amount) => {
+  if (textArr.indexOf('Minimumprovision') > 0) {
+    return +Big(parseGermanNum(textArr[textArr.findIndex(t => t.includes('Minimumprovision')) + 3]));
```
Comment (dredav) [inline src/brokers/commerzbank.js:103]: Also the `indexOf`/`includes` thing.

File: src/brokers/commerzbank.js [hunk 4]
```diff
@@ -105,9 +132,17 @@ const findCompanyDividend = (textArr, foreignDividend = false) => {
...
-  return textArr.slice(startCompanyName, startCompanyName + 2).join(' ');
+  else {
+    if (textArr.indexOf('Investment-Ausschüttung') > 0) {
+      const startCompanyName = textArr.findIndex(t => t.includes('Investment-Ausschüttung')) + 5;
+	  return textArr.slice(startCompanyName, startCompanyName + 2).join(' ');
+	} else if (textArr.indexOf('Inhaber-Anteile') > 0) {
+	  return textArr[textArr.findIndex(t => t.includes('Inhaber-Anteile')) + 1]
+	} else if (textArr.indexOf('Wertpapier-Bezeichnung') > 0) {
+	  const startCompanyName = textArr.findIndex(t => t.includes('Wertpapier-Bezeichnung')) + 5;
+      return textArr.slice(startCompanyName, startCompanyName + 2).join(' ');
```
Comment (dredav) [inline src/brokers/commerzbank.js:143]: Also the `indexOf`/`includes` thing.

File: tests/brokers/commerzbank.test.js [hunk 5]
```diff
@@ -280,6 +299,68 @@ describe('Broker: commerzbank', () => {
...
+
+      expect(result.activities.length).toEqual(1);
+      expect(result.activities[0]).toEqual({
+        broker: 'commerzbank',
+        type: 'Dividend',
+        date: '2019-11-21',
+        datetime: '2019-11-21T' + result.activities[0].datetime.substring(11),
+        isin: 'DE0009848119',
+        wkn: '984811',
+        company: 'LD',
```
Comment (dredav) [inline tests/brokers/commerzbank.test.js:336]: This should be `DWS TOP DIVIDENDE LD`

File: tests/brokers/commerzbank.test.js [hunk 6]
```diff
@@ -280,6 +299,68 @@ describe('Broker: commerzbank', () => {
...
+        date: '2020-12-14',
+        datetime: '2020-12-14T' + result.activities[0].datetime.substring(11),
+        isin: 'US5949181045',
+        wkn: '870747',
+        company: 'Microsoft Corp.',
+        shares: 40,
+        price: 0.39275,
+        amount: 15.71,
+        fee: 0,
+        tax: 0,
```
Comment (dredav) [inline tests/brokers/commerzbank.test.js:319]: The withholding tax is missing here.

File: tests/brokers/commerzbank.test.js [hunk 7]
```diff
@@ -280,6 +299,68 @@ describe('Broker: commerzbank', () => {
...
+        broker: 'commerzbank',
+        type: 'Dividend',
+        date: '2020-12-14',
+        datetime: '2020-12-14T' + result.activities[0].datetime.substring(11),
+        isin: 'US5949181045',
+        wkn: '870747',
+        company: 'Microsoft Corp.',
+        shares: 40,
+        price: 0.39275,
+        amount: 15.71,
```
Comment (dredav) [inline tests/brokers/commerzbank.test.js:317]: The amount shoud be gross including the withholding tax. This also affect the price per share.

File: tests/brokers/commerzbank.test.js [hunk 8]
```diff
@@ -194,6 +194,25 @@ describe('Broker: commerzbank', () => {
...
+
+    test('Can a sell order for 938914 be parsed', () => {
+      const result = commerzbank.parsePages(sellSamples[1]);
+
+      expect(result.activities.length).toEqual(1);
+      expect(result.activities[0]).toEqual({
+        broker: 'commerzbank',
+        type: 'Sell',
+        date: '2020-12-23',
+        datetime: '2020-12-23T' + result.activities[0].datetime.substring(11),
```
Comment (dredav) [inline tests/brokers/commerzbank.test.js:206]: Could you add here the full date static? The time should be 12:36.

File: src/brokers/commerzbank.js [hunk 9]
```diff
@@ -65,19 +79,32 @@ const findIsinDividend = (textArr, foreignDividend) => {
   if (foreignDividend) {
     isinIdx = textArr.findIndex(t => t.includes('STK')) + 2;
   } else {
```
Comment (dredav) [inline src/brokers/commerzbank.js:81]: Minor: I would like to reduce the indents. This can be converted to a `else if`.

**Review by dredav — CHANGES_REQUESTED** [review]

File: src/brokers/commerzbank.js (hunk 1 above)
Comment (oxi0) [reply src/brokers/commerzbank.js:62]: Fixed

File: src/brokers/commerzbank.js (hunk 2 above)
Comment (oxi0) [reply src/brokers/commerzbank.js:84]: Looks better, thanks for the note. 
Fixed

File: src/brokers/commerzbank.js (hunk 3 above)
Comment (oxi0) [reply src/brokers/commerzbank.js:103]: Fixed

File: src/brokers/commerzbank.js (hunk 4 above)
Comment (oxi0) [reply src/brokers/commerzbank.js:143]: Fixed

File: tests/brokers/commerzbank.test.js (hunk 5 above)
Comment (oxi0) [reply tests/brokers/commerzbank.test.js:336]: I had already noticed that, but the document only offers this name. The full name appears to be missing here and only this shortcut is given.

File: tests/brokers/commerzbank.test.js (hunk 6 above)
Comment (oxi0) [reply tests/brokers/commerzbank.test.js:319]: Fixes

File: tests/brokers/commerzbank.test.js (hunk 5 above)
Comment (dredav) [reply tests/brokers/commerzbank.test.js:336]: The name is written some lines above (before the number of shares). Is is possible to have a smart logic to match them also?

File: tests/brokers/commerzbank.test.js (hunk 7 above)
Comment (oxi0) [reply tests/brokers/commerzbank.test.js:317]: This behavior seems to have been implemented this way for a long time. I have now adjusted the logic, but had to change three old test cases. I hope that was okay. The changes looked okay to me.

File: tests/brokers/commerzbank.test.js (hunk 8 above)
Comment (oxi0) [reply tests/brokers/commerzbank.test.js:206]: Fixed

File: src/brokers/commerzbank.js (hunk 9 above)
Comment (oxi0) [reply src/brokers/commerzbank.js:81]: Fixed

File: tests/brokers/commerzbank.test.js (hunk 5 above)
Comment (oxi0) [reply tests/brokers/commerzbank.test.js]: Okay, I missed that. The logic has been adjusted.

File: tests/brokers/commerzbank.test.js (hunk 7 above)
Comment (SirGibihm) [reply tests/brokers/commerzbank.test.js]: Bugs such as these are to be expected in our codebase ^^'
Thanks for fixing it!

**Comment by SirGibihm** [comment]
@oxi0 Looking good, merging! Sorry it took us some time to review. Real life sometimes gets in the way

**Review by SirGibihm — APPROVED** [review]
Looking good, all requests by @dredav have been handled. @dredav Could you mark your changes as resolved if you agree? I can take care of merging

**Comment by oxi0** [comment]
Thank you for accepting the merge request. 
I noticed that the tickets are still open. Can you fix it @SirGibihm or @dredav?

**Comment by dredav** [comment]
Oh yes, I've closed the issues. The issues wasn't closed automatically because the issues wasn't linked to this PR.  /cc @SirGibihm

**Pull request merged**